audacity/audacity
# Time stretching 4 of 6 repair copy paste for stretched tracks

Resolves: #4850 
Resolves: #5273

Copying a clip of a given stretch ratio onto a clip of another stretch ratio cannot work as it has been so far. Instead of rendering the time stretching, we create new clips.

Note that this PR also changes the behavior if clips have equal ratios, since this was an ask from Product (@Tantacrul @LWinterberg ) and a low-hanging fruit for this PR.

Also note that this PR removes the feature whereby clicking the junction of two clips merges them, as a provisional step towards https://github.com/audacity/audacity/issues/2330.




Recommended:

## QA

Please try a myriad of copy, paste and join operations, including trimmed clips, stretched clips, ...
Copy-paste operations should be tried with the "Editing a clip can move other clips" preference both on and off.
- [x] Cross-project copy-paste with difference of project tempo; observe time selection after paste, and stretching of the insertion which plays back at the new tempo
- [x] ~~Pasting a smart clip at the exact end of another clip preserves right-hand hidden data, loses (doesn't paste) left-hand~~ (The behaviour should be kept as in current release, that is, trimmed data should remain recoverable by untrimming).
**Peter 05Oct23 testing with latest 4 of 6 branch:**  The entire smart clip appears to be pasted. I had  30 second chirp and smart clipped it  by 10 seconds each side .  Then pasted it adjacent to  the right end of a 30 second tone.  in situ I could re-expand the chirp on the right.  I could expand it on the left in situ, but if I moved it rightwards away from the  tone I could re-expand it on the left too.  I absolutely think this is correct/desired behavior, the entire smart clip should be retained,
- [x] ~~Vice-versa, when pasting at the exact start of another clip~~
**Peter 05Oct23 testing with latest 4 of 6 branch:**  similarly so
- [x] Join stretched audio clips with equal stretch ratios doesn't render stretching
**Peter 05Oct23 testing with latest 4 of 6 branch:**  it still does render to 100%
- [x] Join audio clips with different stretch ratios shows progress bar
 **Peter 05Oct23 testing with latest 4 of 6 branch:**  Progress bar is properly shown
- [x] `Edit > Labeled Audio > Join` on project with several tracks of varying stretch ratios. (#5246)
 **Peter 05Oct23 testing with latest 4 of 6 branch:**  Can't test as labelled audio commands are FUBAR
- [x] Likewise with another label-based editing command of your choice, such as `Labeled Audio > Delete`.
 **Peter 05Oct23 testing with latest 4 of 6 branch:**  Can't test as labelled audio commands are FUBAR
- [x] Sync-lock when a wave track with stretch lengthens, inserting silence into a clip
- [x] Generating audio over a stretched clip yields expected result, be it silence, an Audacity or a Nyquist tone.
**Peter 05Oct23 testing with latest 4 of 6 branch:**  This looks to work properly in the branch - the generated audio is 100%, but the audio clips either side retain their stretch % factor (and are not rendered).  Expected and desired behavior I think.
- [x] Envelope still working, also when drawing. Please make sure the end samples are correctly scaled and editable.
- [x] Fixes https://github.com/audacity/audacity/issues/4984 and https://github.com/audacity/audacity/issues/5145 (please close if it does)
**Peter 05Oct23 testing with latest 4 of 6 branch:**  note the cliplet/"invisible" clip syndrome.
- [x] Any chance it fixes https://github.com/audacity/audacity/issues/4297 ?
**Peter 05Oct23 testing with latest 4 of 6 branch:**    Looks to be fixed in the 4 of 6 branch
- [x] Paste a clip A onto a portion of a clip B. Now move A out of the way. The data in B that was under A is recoverable by un-trimming.
**Peter 05Oct23 testing with latest 4 of 6 branch:**    I can't see how to do the paste-over, so can't test this.
- [x] "Silence Audio" command over several tracks and stretched clips shows a convincing progress bar
**Peter 05Oct23 testing with latest 4 of 6 branch:**   I get a "convincing progress bar" with both stretched and unstretched clips - each clip was a one hour chirp and I had four of them.  The silent audio was properly rendered to 100%
- [x] Same for the AutoDuck effect, with stretch in the control track.
**Peter 05Oct23 testing with latest 4 of 6 branch:**   I get a "convincing progress bar" with both stretched and unstretched clips - each clip was a one hour chirp and I had four of them.  The ducked audio was properly rendered to 100%
- [x] Equalization effects (such as Graphic EQ) on unstretched tracks have the same behavior as before (i.e., merging begin and end of output when possible, preserving clip boundaries within the selection, etc.)
**Peter 05Oct23 testing with latest 4 of 6 branch:**   Seems to work properly. But if I have a gap between clips I get an unwanted additional silent clip between the separated clips (which of course impacts on clip movements)
- [x] Verify that this PR fixes #5265
**Peter 05Oct23 testing with latest 4 of 6 branch:**    This looks to work properly in the branch
- [x] Check if this PR fixes #4250
**Peter 05Oct23 testing with latest 4 of 6 branch:**    This looks to work properly in the branch
- [x] [Bug found by @Tantacrul on master](https://www.dropbox.com/s/0hlb011cx0rx4wr/Pasting%20Mutliple%20clips%20next%20to%20other%20clips.mov?dl=0) is fixed
- [x] Implements #5273
**Peter 05Oct23 testing with latest 4 of 6 branch:**    This renders the clips but I get glitches/clicks - see bug thread for details
- [x] Insert silence on stretched clip with non-trivial envelope behaves as expected
- [x] Cutline-cutting stretched region with envelope and expanding back restores the envelope
- [x] Paulstretch on both unstretched and stretched clip
**Peter 05Oct23 testing with latest 4 of 6 branch:**    seems to work fine with both stretched and un-stretched.
